Split Session class into interface/implementation parts

glassez opened the pull request:
This saves from deep recompilation with every minor change in the implementation of the Session class affecting its header file, since a lot of code depends on it.

thalieht commented:
Maybe you should consider start using more than 1 commits for such big PRs... like at least one commit to rename `session` to `sessionImpl` and the rest up to you. Right now i'm thankful i can't really review code.

glassez commented:
@thalieht 
For some reason, I think of commits as something integral, representing a separate completed task, after which the program remains compilable and workable (apparently, the habit of seeing a clean history). But you're right, it can be unfriendly to review. In fact, it's not such a problem to push a few intermediate commits and then squash them before merging. It is more problematic to change my habit...

>at least one commit to rename `session` to `sessionImpl` and the rest up to you. 

I don't know if I can change it retroactively, but I'll try.

Chocobo1 commented:
Why not define an interface instead?

glassez commented:
> Why not define an interface instead? Compared to an interface, PIMPL seems like a workaround for certain problems of the language/compiler IMO.

Well, in a case like this, both options are used as "a workaround for certain problems of the language". I was thinking that PIMPL could have advantages over the interface... But it seems that these advantages do not play any role in our case (as opposed to developing libraries), so I changed my mind and switched to using a public interface instead.

I tried to split it into a few intermediate commits to make it easier to review. They will be squashed when merging.

glassez commented:
@qbittorrent/bug-handlers
Sorry for offtopic. Can you take a look at https://github.com/arvidn/libtorrent/pull/7025? (I'm not sure I can mention qBittorrent teams from repository of another organization.)

thalieht commented:
No problems found.

Chocobo1 commented on src/base/bittorrent/CMakeLists.txt:
```diff
@@ -0,0 +1,225 @@
+add_library(qbt_base STATIC
```
Doesn't seem correct to add this file.

glassez replied:
>Doesn't seem correct to add this file.

Apparently, I accidentally copied them to this folder.

Chocobo1 commented on src/base/bittorrent/base.pri:
```diff
@@ -0,0 +1,188 @@
+HEADERS += \
```
This too.

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -0,0 +1,486 @@
+
+#pragma once
+
+//#include <QtContainerFwd>
```
Is it leftover?

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -0,0 +1,486 @@
+#pragma once
+
+//#include <QtContainerFwd>
+#include <QHash>
```
Seems it is not really in use, so you can use `#include <QtContainerFwd>` instead.

glassez replied:
All done.

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -177,90 +151,45 @@ namespace BitTorrent
+    class Session : public QObject
     {
         Q_OBJECT
-        Q_DISABLE_COPY_MOVE(Session)
```
Is it required to remove it?

glassez replied:
>Is it required to remove it?

No.

Chocobo1 commented on src/base/bittorrent/session.h:
```diff
@@ -29,51 +29,21 @@
-#include "base/settingvalue.h"
-#include "base/types.h"
+#include "base/pathfwd.h"
 #include "addtorrentparams.h"
 #include "cachestatus.h"
 #include "categoryoptions.h"
 #include "sessionstatus.h"
 #include "torrentinfo.h"
 #include "trackerentry.h"
```
I think some of those headers can be forward declared.

Chocobo1 replied:
Or maybe they are all required...

glassez replied:
This will most likely affect many sources that include this header. I didn't intend to do it since it isn't required for the job. It can be done later.

Chocobo1 commented on src/base/bittorrent/sessionimpl.cpp:
```diff
@@ -320,15 +320,55 @@ struct BitTorrent::Session::ResumeSessionContext final : public QObject
 
 const int addTorrentParamsId = qRegisterMetaType<AddTorrentParams>();
 
-// Session
```
I don't insist but I think it will help to have this comment and `// SessionImpl` to show where are the boundaries of code segments of each class.
